core/src/trezor/crypto/slip39.py

```diff
@@ -0,0 +1,592 @@
+# Copyright (c) 2018 Andrew R. Kozlik
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy of
+# this software and associated documentation files (the "Software"), to deal in
+# the Software without restriction, including without limitation the rights to
+# use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
+# of the Software, and to permit persons to whom the Software is furnished to do
+# so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in all
+# copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
+# WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+# CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+#
+
+import math
+
+from trezor.crypto import hashlib, hmac, pbkdf2, random
+from trezor.crypto.slip39_wordlist import wordlist
+from trezorcrypto import shamir
+
+
+class ConfigurationError(Exception):
+    pass
+
+
+class MnemonicError(Exception):
+    pass
+
+
+RADIX_BITS = 10
+"""The length of the radix in bits."""
+
+RADIX = 2 ** RADIX_BITS
+"""The number of words in the wordlist."""
+
+ID_LENGTH_BITS = 15
+"""The length of the random identifier in bits."""
+
+ITERATION_EXP_LENGTH_BITS = 5
+"""The length of the iteration exponent in bits."""
+
+ID_EXP_LENGTH_WORDS = (ID_LENGTH_BITS + ITERATION_EXP_LENGTH_BITS) // RADIX_BITS
+"""The length of the random identifier and iteration exponent in words."""
+
+MAX_SHARE_COUNT = 16
+"""The maximum number of shares that can be created."""
+
+CHECKSUM_LENGTH_WORDS = 3
+"""The length of the RS1024 checksum in words."""
+
+DIGEST_LENGTH_BYTES = 4
+"""The length of the digest of the shared secret in bytes."""
+
+CUSTOMIZATION_STRING = b"shamir"
+"""The customization string used in the RS1024 checksum and in the PBKDF2 salt."""
+
+METADATA_LENGTH_WORDS = ID_EXP_LENGTH_WORDS + 2 + CHECKSUM_LENGTH_WORDS
+"""The length of the mnemonic in words without the share value."""
+
+MIN_STRENGTH_BITS = 128
+"""The minimum allowed entropy of the master secret."""
+
+MIN_MNEMONIC_LENGTH_WORDS = METADATA_LENGTH_WORDS + math.ceil(MIN_STRENGTH_BITS / 10)
```

Comment:
I prefer using `(MIN_STRENGTH_BITS + 9) // 10` although it might be a little bit cryptical. There are two good reasons for that: We'll avoid using floating point numbers and we'll drop the `math` dependency.

Reply:
I agree, but I defined bits_to_bytes() and bits_to_words() for better legibility.